File: simnibs/examples/simulations/tACSchallenge.py
"""Script for simulating the TACS challenge montage.

see @TACSchallenge on Twitter

A. Thielscher, 2020
"""
from copy import deepcopy
from simnibs import sim_struct, run_simnibs


### general Information
S = sim_struct.SESSION()
S.subpath = 'm2m_ernie'  # head mesh
S.pathfem = 'TACSchallenge'  # Directory for the simulation
S.map_to_surf=True
S.open_in_gmsh=True # open results once they are ready
                    # (set to False if you are annoyed by the popup windows)

### Define Condition A
ConditionA = S.add_tdcslist()

### Set current flow though each channel:
### 2mA peak-to-peak --> 1 mA baseline-to-peak
### The third entry is a "pseudochannel" to which electrodes 
### are connected that are not used in the simulated condition
ConditionA.currents = [0.001, -0.001, 0.0]

# Define occipital electrodes
O1 = ConditionA.add_electrode()
O1.channelnr = 1  # Connect the electrode to the first channel
O1.centre = 'O1'  # Place it over O1
O1.shape = 'ellipse'  # Elliptical / circular shape
O1.dimensions = [25, 25]  # Electrode diameter (in mm)
O1.thickness = [1.5, 1.5]  # 1.5 mm gel and 1.5 mm rubber 

ConditionA.electrode.append(deepcopy(O1))
ConditionA.electrode[-1].centre = 'Oz'

ConditionA.electrode.append(deepcopy(O1))
ConditionA.electrode[-1].centre = 'O2'

# Define return electrodes
CP5 = ConditionA.add_electrode()
CP5.channelnr = 2;
CP5.centre = 'CP5';
CP5.shape = 'ellipse';
CP5.dimensions = [50, 50];
CP5.thickness = [1.5, 1.5];

ConditionA.electrode.append(deepcopy(CP5))
ConditionA.electrode[-1].centre = 'CPz'

ConditionA.electrode.append(deepcopy(CP5))
ConditionA.electrode[-1].centre = 'CP6'

# Define retinal control electrodes
Retcontr = ConditionA.add_electrode()
Retcontr.channelnr = 3; # connect to the "pseudochannel"
Retcontr.centre = [35.4, 106.0, -40.6]; # coordinates determined in simnibs_gui
Retcontr.shape = 'ellipse';
Retcontr.dimensions = [25, 25];
Retcontr.thickness = [1.5, 1.5];

ConditionA.electrode.append(deepcopy(Retcontr))
ConditionA.electrode[-1].centre = [-33.2, 108, -40.6]; # coordinates determined in simnibs_gui


# Define Condition B
S.poslists.append(deepcopy(ConditionA))
ConditionB=S.poslists[-1]

ConditionB.electrode[0].channelnr = 3;
ConditionB.electrode[1].channelnr = 3;
ConditionB.electrode[2].channelnr = 3;
ConditionB.electrode[6].channelnr = 1;
ConditionB.electrode[7].channelnr = 1;


# Define Condition C
S.poslists.append(deepcopy(ConditionA))
ConditionC=S.poslists[-1]

ConditionC.electrode[0].channelnr = 2;
ConditionC.electrode[1].channelnr = 1;
ConditionC.electrode[2].channelnr = 2;
ConditionC.electrode[3].channelnr = 3;
ConditionC.electrode[4].channelnr = 3;
ConditionC.electrode[5].channelnr = 3;
ConditionC.electrode[6].channelnr = 3;
ConditionC.electrode[7].channelnr = 3;

run_simnibs(S)

File: simnibs/examples/simulations/TACSchallenge.m
%
% script for simulating the TACS challenge montage
% see @TACSchallenge on Twitter
% 
% A. Thielscher, 2020
%

%% General information
S = sim_struct('SESSION');
S.subpath = 'm2m_ernie'; % subject folder
S.pathfem   = 'TACSchallenge'; % folder for the simulation output
S.map_to_surf = true;
S.open_in_gmsh = true; % open results once they are ready
                       % (set to false if you are annoyed by the popup windows)

                       
%% Define Condition A

% Set current flow though each channel:
% 2mA peak-to-peak --> 1 mA baseline-to-peak
% The third entry is a "pseudochannel" to which electrodes 
% are connected that are not used in the simulated condition
S.poslist{1} = sim_struct('TDCSLIST');
S.poslist{1}.currents = [0.001, -0.001, 0.0]; 

% Define occipital electrodes
S.poslist{1}.electrode(1).channelnr = 1; % Connect the electrode to the first channel
S.poslist{1}.electrode(1).centre = 'O1'; % Place it over O1
S.poslist{1}.electrode(1).shape = 'ellipse'; % Define it to be elliptical/circular
S.poslist{1}.electrode(1).dimensions = [25, 25]; % Electrode diameter (in mm)
S.poslist{1}.electrode(1).thickness = [1.5, 1.5]; % 1.5 mm gel and 1.5 mm rubber 

S.poslist{1}.electrode(2) = S.poslist{1}.electrode(1);
S.poslist{1}.electrode(2).centre = 'Oz';

S.poslist{1}.electrode(3) = S.poslist{1}.electrode(1);
S.poslist{1}.electrode(3).centre = 'O2';

% Define return electrodes
S.poslist{1}.electrode(4).channelnr = 2; % Connect to the second channel
S.poslist{1}.electrode(4).centre = 'CP5';
S.poslist{1}.electrode(4).shape = 'ellipse';
S.poslist{1}.electrode(4).dimensions = [50, 50];
S.poslist{1}.electrode(4).thickness = [1.5, 1.5];

S.poslist{1}.electrode(5) = S.poslist{1}.electrode(4);
S.poslist{1}.electrode(5).centre = 'CPz';

S.poslist{1}.electrode(6) = S.poslist{1}.electrode(4);
S.poslist{1}.electrode(6).centre = 'CP6';

% Define retinal control electrodes
S.poslist{1}.electrode(7).channelnr = 3; % Connect to the "pseudochannel"
S.poslist{1}.electrode(7).centre = [35.4, 106.0, -40.6]; % coordinates determined in simnibs_gui
S.poslist{1}.electrode(7).shape = 'ellipse';
S.poslist{1}.electrode(7).dimensions = [25, 25];
S.poslist{1}.electrode(7).thickness = [1.5, 1.5];

S.poslist{1}.electrode(8) = S.poslist{1}.electrode(7);
S.poslist{1}.electrode(8).centre = [-33.2, 108, -40.6]; % coordinates determined in simnibs_gui


%% Define Condition B (retinal control)
S.poslist{2} = S.poslist{1};

S.poslist{2}.electrode(1).channelnr = 3;
S.poslist{2}.electrode(2).channelnr = 3;
S.poslist{2}.electrode(3).channelnr = 3;
S.poslist{2}.electrode(7).channelnr = 1;
S.poslist{2}.electrode(8).channelnr = 1;


%% Define Condition C (cutaneous control)
S.poslist{3} = S.poslist{1};

S.poslist{3}.electrode(1).channelnr = 2;
S.poslist{3}.electrode(2).channelnr = 1;
S.poslist{3}.electrode(3).channelnr = 2;
S.poslist{3}.electrode(4).channelnr = 3;
S.poslist{3}.electrode(5).channelnr = 3;
S.poslist{3}.electrode(6).channelnr = 3;
S.poslist{3}.electrode(7).channelnr = 3;
S.poslist{3}.electrode(8).channelnr = 3;


%% Run Simulation
run_simnibs(S);
